src/anemoi/inference/clusters/cluster.py

```diff
@@ -0,0 +1,183 @@
+# (C) Copyright 2025- ECMWF.
+#
+# This software is licensed under the terms of the Apache Licence Version 2.0
+# which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
+# In applying this licence, ECMWF does not waive the privileges and immunities
+# granted to it by virtue of its status as an intergovernmental organisation
+# nor does it submit to any jurisdiction.
+#
+
+import datetime
+import logging
+import os
+from abc import ABC
+from abc import abstractmethod
+from typing import Any
+from typing import NamedTuple
+
+import numpy as np
+
+from anemoi.inference.context import Context
+from anemoi.inference.lazy import torch
+
+ADDRESS = NamedTuple("Address", [("host", str), ("port", int)])
+LOG = logging.getLogger(__name__)
+
+
+class Cluster(ABC):
+    """Abstract base class for cluster and parallel environment handling."""
+
+    _model_comm_group: torch.distributed.ProcessGroup | None = None  # type: ignore
+
+    def __init__(self, context: Context, **kwargs: Any) -> None:
+        """Cluster class for parallel inference
+
+        Parameters
+        ----------
+        context : Context
+            Runner context
+        kwargs : Any
+            Additional keyword arguments
+        """
+        self.context = context
+        _ = kwargs  # To avoid unused variable warning
+
+    @classmethod
+    @abstractmethod
+    def used(cls) -> bool:
+        """Check if this cluster is valid in the current environment."""
+        pass
+
+    def spawn(self, fn: Any, *args: Any) -> None:
+        """Spawn processes in the cluster environment.
```

Comment:
To check I understand the intention. Let's say we want a Cascade-based cluster, which already has gang capability. This means we would launch a gang, and each of the tasks would get world size and rank and coordinator address

what is each task's callable then, from the PoV of this class? I would expect something that calls the constructor of this class to populate the master address/port and world size and rank from the envvars, and call initialise. And then continue with the actual torch code

but is there a place for spawn then? Is spawn always expected to be invoked, or is it only to be so for "manual" clusters?